File: day2/solution.py
import os 
input_file = open(os.getcwd() + '/input.txt', 'r').read().split(',')
sample_input_file = open(os.getcwd() + '/sample_input.txt', 'r').read().split(',')
sample_array = ["11-22"]

def check_invalid_id_2(id):
    for window_len in range(1, len(id)):
        for window_start in range(len(id)):
            if window_start + window_len * 2 > len(id):
                break

            #print(f"Winstart: {window_start} Winlen: {window_len}")
            #print(f"Winstart: {window_start + window_len} Winlen: {window_start + window_len * 2}")
            substring1 = id[window_start:window_start+window_len]
            substring2 = id[window_start+window_len:window_start + window_len * 2]

            #print(f"{substring1} and {substring2}")
            if substring1 == substring2:
                return True
            
    return False

def get_invalid_id_2(input_file):

    invalid_id_array = []

    for id_range in input_file:
        first_id = int(id_range.split("-")[0])
        last_id = int(id_range.split("-")[1])
        invalid_ids = []

        for id in range(first_id, last_id+1):
            print(f"ID {id}, ID Length: {len(str(id))}")
            if check_invalid_id_2(str(id)):
                invalid_ids.append(id)

        if len(invalid_ids) > 0:
            print(f"{id_range} has {len(invalid_ids)} invalid ID(s), {invalid_ids}")
        invalid_id_array += invalid_ids
            
    print("The rest of the ranges contain no invalid IDs.")

    invalid_id_array = [int(i) for i in invalid_id_array]

    print(f"Invalid IDs: {invalid_id_array}. ")
    print(f"Adding up all the invalid IDs produces {sum(invalid_id_array)}.")


def get_invalid_id_1(input_file):

    invalid_id_array = []

    for id_range in input_file:
        first_id = int(id_range.split("-")[0])
        last_id = int(id_range.split("-")[1])
        invalid_ids = []

        for id in range(first_id, last_id+1):
            id_string = str(id)
            #print(f"ID {id}, ID Length: {len(str(id))}")
            #print(id_string[0:len(id_string)//2], id_string[len(id_string)//2: len(id_string)])
            if id_string[0:len(id_string)//2] == id_string[len(id_string)//2: len(id_string)]:
                invalid_ids.append(id_string)

        if len(invalid_ids) > 0:
            print(f"{id_range} has {len(invalid_ids)} invalid ID(s), {invalid_ids}")
        invalid_id_array += invalid_ids
            
    print("The rest of the ranges contain no invalid IDs.")

    invalid_id_array = [int(i) for i in invalid_id_array]

    print(f"Invalid IDs: {invalid_id_array}. ")
    print(f"Adding up all the invalid IDs produces {sum(invalid_id_array)}.")

#get_invalid_id_1(sample_input_file)
#get_invalid_id_1(input_file)

get_invalid_id_2(sample_input_file)           

File: day2/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <numeric>

using namespace std;

vector<string> split_string(string input_string, char delimiter)
{
    stringstream input_stream(input_string);
    string segment;
    vector<string> seglist;

    while (getline(input_stream, segment, delimiter))
    {
        seglist.push_back(segment);
    }

    return seglist;
}

void get_invalid_id_1(string input_file)
{

    ifstream infile(input_file);
    stringstream buffer;

    buffer << infile.rdbuf();
    string line = buffer.str();

    infile.close();

    stringstream id_ranges(line);
    string id_range;

    vector<long long> invalid_id_vector;

    while (getline(id_ranges, id_range, ','))
    {
        //cout << id_range << '\n';
        vector<string> ids = split_string(id_range, '-');
        string first_id = ids.front();
        string last_id = ids.back();

        cout << "Last ID: " + last_id << '\n';

        vector<long long> invalid_ids;

        for (long long id = stoll(first_id); id <= stoll(last_id); id++) {
            string id_string = to_string(id);
            /*cout << "ID: " + id_string 
            + " ID string length: " 
            + to_string(id_string.length()) << '\n';*/

            if (id_string.substr(0, id_string.length()/2)
            .compare(id_string.substr(id_string.length()/2, 
            id_string.length() - id_string.length()/2)) == 0) {
                invalid_ids.push_back(id);
            }
        }

        if (invalid_ids.size() > 0) {
            cout << id_range + " has " +
            to_string(invalid_ids.size()) + 
            " invalid IDs, ";
            for (long long i = 0; i < invalid_ids.size(); i++) {
                if (i > 0) {
                    cout << ", ";
                }
                cout << invalid_ids[i];
            }
            cout << '\n';
        }

        invalid_id_vector.insert(invalid_id_vector.end(),
        invalid_ids.begin(), invalid_ids.end());
    }
    cout << "The rest of the ranges contain no invalid IDs." << '\n';

    long long sum_ids = accumulate(invalid_id_vector.begin(),
    invalid_id_vector.end(), 0LL);

    cout << "Adding up all the invalid IDs produces "
    + to_string(sum_ids) << '\n';

}

int main()
{
    string sample_input = "sample_input.txt";
    string input = "input.txt";
    string test_input = "test_input.txt";

    //get_invalid_id_1(sample_input);
    get_invalid_id_1(input);
}